### PR TITLE
Upgrade to stable constraints

### PR DESCRIPTION
**Pull request by potiuk**
This change upgrades setup.py to provide non-conflicting
`pip check` valid set of constraints for CI image.

PIP check is run once the images are available.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
@turbaszek - I moved your changes to "Wait for images" step. As soon as #12635 gets merged, this PR (with "upgrade to latest dependencies" label added) should (I believe) cleanly build and successfully run pip-check (at least for the CI Images). Once the constraints are updated, we can also likely enable prod image check.

**Comment by potiuk** [comment]
I can split it further into two PRs, but I want to make it pass and see if indeed PIP check works @turbaszek !

**Review by turbaszek — APPROVED** [review]

**Comment by potiuk** [comment]
I still disable the result of the pip check - they will be printed but won't fail the build. Working on making sure all those are solved

**Comment by potiuk** [comment]
@turbaszek -> final (????) test.

**Comment by potiuk** [comment]
ERROR    airflow.models.dagbag.DagBag:dagbag.py:297 Failed to import: /opt/airflow/airflow/providers/google/ads/example_dags/example_ads.py
Traceback (most recent call last):
  File "/opt/airflow/airflow/models/dagbag.py", line 294, in _load_modules_from_file
    loader.exec_module(new_module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/opt/airflow/airflow/providers/google/ads/example_dags/example_ads.py", line 24, in <module>
    from airflow.providers.google.ads.operators.ads import GoogleAdsListAccountsOperator
  File "/opt/airflow/airflow/providers/google/ads/operators/ads.py", line 24, in <module>
    from airflow.providers.google.ads.hooks.ads import GoogleAdsHook
  File "/opt/airflow/airflow/providers/google/ads/hooks/ads.py", line 25, in <module>
    from google.ads.google_ads.v2.types import GoogleAdsRow
ModuleNotFoundError: No module named 'google.ads.google_ads.v2'

File: setup.py
Comment (turbaszek) [inline setup.py:250]:
```suggestion
    'google-ads>=3.3.0,<8.0.0',
```
Should support 3.6 and it has v2 API

Comment (ashb) [reply setup.py:250]: Does 4.0 drop support for Py 3.6?

If so we'll need a more complex thing here including `; python_version...`

Comment (turbaszek) [reply setup.py:250]: > If so we'll need a more complex thing here including

Is it necessary? Won't pip resolver use the right version for each python?

Comment (ashb) [reply setup.py:250]: Ahahahahahaah.

No, don't think so :( I'd like to be wrong here though.

Comment (potiuk) [reply setup.py:250]: The 8.0.0 limitation is needed, because the Ads operators are using v2 version of the API and support for those have been removed in 8.0.0 (very recently - in November). The operator should be updated (but not now).

The 4.0.0 still supports 3.6.7+. Support for 3.7 has been removed in one of the 4.1.* versions. PYPI solves it nicely on their own as those non-3.7 packages have the limitation to Python>=3.7 set and pypi will not go beyond that version in 3.6.

Comment (potiuk) [reply setup.py:250]: See here: 4.0.0 still supports 3.6.7+  https://pypi.org/project/google-ads/4.0.0/

setup.py:
```
    python_requires='>=3.6',
```

Comment (ashb) [reply setup.py:250]: Oh I'm confusing the case where we don't want the module for a python version at all. Sorry!

**Comment by kaxil** [comment]
This is ready for review / merge right? Or are you waiting on something @potiuk ?

**Review by turbaszek — APPROVED** [review]

**Comment by potiuk** [comment]
> This is ready for review / merge right? Or are you waiting on something @potiuk ?

I am looking at the result of it now and try to figure out the final version of it. I still see that even if we remove/install the packages, pip will - for example install request == 2.25.1 even if we explicitly stated `requests< 2.24.0`. I already figured out god set of constraints for Py3.6 and I now have to get it also for 3.7 and 3.8. 

Also - I temporarily switched to`constraints-pipcheck` branch and once we merge it and update the master constraints to match I will have to reverse it. So few more steps are needed.

**Comment by potiuk** [comment]
OK. I figured out EXACTLY what the problem with eager upgrade strategy was and wy we had conflicting requirements at the end of it.

PIP first figures out the requirements for airflow itself and after that it adds extra.

So when it first tries to get the requirements to be installed it installs them for the (`install_requires` in setup.cfg). And whatever is installed there, will not be downgraded even if a lower version is required by one of the extras. 

This happened with `requests` - airflow "install_requires" figures that the best version to install is 2.25.0. But when snowlake and other deps have < 2.24.0 - requests will not be downgraded to 2.23.0, even if this produces version conflict. 

That's why "eager" upgrade strategy upgraded requests to 2.25.0 -> because "install_requires" deps told it to do so. 

The solution to all such conflicting requirements is simply to limit them in "install_requires" section :D. Then --upgrade-staregy eager strategy just works as expected.

**Comment by ashb** [comment]
Nice find.

"Shocked" to discover that pip doesn't have an actual dependency resolver, but some random code instead :)

**Comment by potiuk** [comment]
> Nice find.
> 
> "Shocked" to discover that pip doesn't have an actual dependency resolver, but some random code instead :)

I am also going to try the new resolver now, when I have good set of instal/extra requirements.

The nice thing about it, is that likely we will not have to do any "tricks" with pipcheck constraints. Once I put the right requirements in install_requires, it seems that we will get "upgraded" pip-check compliant constraints

**Comment by potiuk** [comment]
3.8: docker run --rm --entrypoint /bin/bash apache/airflow:master-python3.8-ci -c pip check
No broken requirements found.
3:7: docker run --rm --entrypoint /bin/bash apache/airflow:master-python3.7-ci -c pip check
No broken requirements found.
3.6: docker run --rm --entrypoint /bin/bash apache/airflow:master-python3.6-ci -c pip check
No broken requirements found.

**Comment by potiuk** [comment]
Now - there is only one import problem with snowflake, but it should be an easy fix!

**Comment by potiuk** [comment]
I am about to merge this one @kaxil @ash - if you want to take a  look. I xfailed the WasbHook tests and handle import failure for now. This should not stop teh WasbHook from working (it just cannot be tested when snowflake is installed) and @ephraimbuddy works on the #12188. After this is merged, we will restore proper tests/imports.

I still have to make 'pip check' failure fail the build, but I will add separate PR for that once I merge this one and push updated constraints - I believe once the new constraints will be pushed, also production image will pass the 'pip check'.

**Comment by potiuk** [comment]
OK. Just pushed the new constraints, So I merge this one and let it build.

**Pull request merged**
